server/api/api.go

```diff
@@ -0,0 +1,129 @@
+package api
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/mattermost/mattermost-plugin-autolink/server/link"
+)
+
+const PluginIDContextValue = "pluginid"
+
+type LinkStore interface {
+	GetLinks() []link.Link
+	SaveLinks([]link.Link) error
+}
+
+type Authorization interface {
+	IsAuthorizedAdmin(userID string) (bool, error)
+}
+
+type Handler struct {
+	root          *mux.Router
+	linkStore     LinkStore
+	authorization Authorization
+}
+
+func NewHandler(linkStore LinkStore, authorization Authorization) *Handler {
+	h := &Handler{
+		linkStore:     linkStore,
+		authorization: authorization,
+	}
+
+	root := mux.NewRouter()
+	api := root.PathPrefix("/api/v1").Subrouter()
+	api.Use(h.adminOrPluginRequired)
```

Comment:
@levb this is why I think I'm starting to like gorillamux. The tradeoff is more complexity, but you get to do stuff like this.

Reply:
yy, I am a proxy developer, I _like_ proxies. If anything, my problem with `gorilla.mux` is its lack of better proxy features, like extending the request object with vars, and a handy library of filters.

I do _not_ at all advocate a homegrown solution, I appreciate the benefits of a good `router`, but `gorilla.mux` gives me a distaste. I'm slowly accepting it though.